src/main/java/graphql/schema/PropertyDataFetcherHelper.java

```diff
@@ -42,49 +52,52 @@ public static Object getPropertyValue(String propertyName, Object object, GraphQ
             return ((Map<?, ?>) object).get(propertyName);
         }
 
-        String key = mkKey(object, propertyName);
-        //
-        // if we have tried all strategies before and they have all failed then we negatively cache
-        // the key and assume that its never going to turn up.  This shortcuts the property lookup
-        // in systems where there was a `foo` graphql property but they never provided an POJO
-        // version of `foo`.
-        if (isNegativelyCached(key)) {
-            return null;
-        }
-        // lets try positive cache mechanisms next.  If we have seen the method or field before
+        String cacheKey = mkKey(object, propertyName);
+        // lets try positive cache mechanisms first.  If we have seen the method or field before
         // then we invoke it directly without burning any cycles doing reflection.
-        Method cachedMethod = METHOD_CACHE.get(key);
+        CachedMethod cachedMethod = METHOD_CACHE.get(cacheKey);
         if (cachedMethod != null) {
-            MethodFinder methodFinder = (aClass, methodName) -> cachedMethod;
             try {
-                return getPropertyViaGetterMethod(object, propertyName, graphQLType, methodFinder, environment);
+                return invokeMethod(object, environment, cachedMethod.method, cachedMethod.takesDataFetcherEnvironmentAsOnlyArgument);
             } catch (NoSuchMethodException ignored) {
-                assertShouldNeverHappen("A method cached as '%s' is no longer available??", key);
+                assertShouldNeverHappen("A method cached as '%s' is no longer available??", cacheKey);
             }
         }
-        Field cachedField = FIELD_CACHE.get(key);
+        Field cachedField = FIELD_CACHE.get(cacheKey);
         if (cachedField != null) {
-            try {
-                return getPropertyViaFieldAccess(object, propertyName);
-            } catch (FastNoSuchMethodException ignored) {
-                assertShouldNeverHappen("A field cached as '%s' is no longer available??", key);
-            }
+            return invokeField(object, cachedField);
         }
 
+        //
+        // if we have tried all strategies before and they have all failed then we negatively cache
+        // the cacheKey and assume that its never going to turn up.  This shortcuts the property lookup
+        // in systems where there was a `foo` graphql property but they never provided an POJO
+        // version of `foo`.
+        //
+        // we do this second because we believe in the positive cached version will mostly prevail
+        // but if we then look it up and negatively cache it then lest do that look up next
+        //
+        if (isNegativelyCached(cacheKey)) {
+            return null;
+        }
```

Comment:
I flipped the negative cache down in the order because I think the "positive case" is more likely - that is for any property X on a source object we have a method to get it.

So now if we DONT have a cached method we will try to see if its negatively cached (could not be found before) and the do the full search for it

This is a minor improvement because it avoids a map lookup in the case where we have methods, which I beleive is the best case to chase first